feat: support cargo/runtime dependencies

mrcjkb opened the pull request:
Fixes #126.
Fixes #452.
Fixes #455.

If a Rust project has dependencies, cargo-clippy and cargo check will fail in the pre-commit-check derivation.

This adds a `settings.rust.check.cargoDeps` ~and `settings.runtimeDeps`~ option, which can be used to specify the dependencies, e.g.

```nix
settings = {
  rust.check.cargoDeps = pkgs.rustPlatform.importCargoLock {
    lockFile = ./Cargo.lock;
  };
}
```

and reuses the hooks' `extraPackages` as `nativeBuildInputs` in the derivation, as they are likely to be used in the checks in addition to the devShell environment.

domenkozar commented:
Sorry for the mess, we revamed module structure in https://github.com/cachix/pre-commit-hooks.nix/pull/397, could you port these changes over?

mrcjkb commented:
> Sorry for the mess, we revamed module structure in #397, could you port these changes over?

Done :smile:

mrcjkb commented:
@domenkozar looks like this is conflicted again.

Please let me know if you would like to get this merged, and I will resolve conflicts one more time 😃

domenkozar commented:
Yeah sorry :(

domenkozar commented:
Does https://github.com/cachix/git-hooks.nix/pull/430 address this?

mrcjkb commented:
> Does #430 address this?

It looks like it addresses this partially. 

- `settings.runtimeDeps` seems to be addressed. 
- `settings.rust.cargoDeps` (required e.g. to run clippy in sandboxed/offline mode) isn't, afaict.

I'll have to look into it at a later time.

sandydoo commented:
I'm just wondering if this is the right place to inject these dependencies. Can they go into the packages for `clippy` and `cargo check`?

mrcjkb commented:
> I'm just wondering if this is the right place to inject these dependencies. Can they go into the packages for `clippy` and `cargo check`?

You mean using something like `symlinkJoin`?
I don't think so. They need to be in `nativeBuildInputs` and `cargoSetupHook` is needed so that clippy can find the crates.

> Otherwise, aren't the individual hook `entry`s broken unless executed within this `run` derivation? 

No, they're not. Clippy will fetch the crates from the crate registy (just as `cargo build` does).

> Will a manual `pre-commit run <hook_id>` work inside the shell?

Yes.

sandydoo commented:
> You mean using something like symlinkJoin?

No, I meant using `buildInputs` and friends in a derivation that wraps clippy. But since this is a purely flake check issue, that would make setting up the shell more expensive than it needs to be.

My current concerns/thoughts/ideas:
1. The new options only affect `nix flake check` and that isn't clear from the way they're defined.
2. The new options are essentially `mkDerivation` options. Can and should we generalize this? Is there a design that's basically "give us an attrset and we'll merge it into `mkDerivation`"?
3. Should `runtimeDeps` be added to the shell as well? In that case, can we re-use `extraPackages`?

mrcjkb commented:
I see.

> My current concerns/thoughts/ideas:

Here are my thoughts:

> 1. The new options only affect `nix flake check` and that isn't clear from the way they're defined.

Good point. I guess they could go in a `settings.check` set?

> 2. The new options are essentially `mkDerivation` options. Can and should we generalize this? Is there a design that's basically "give us an attrset and we'll merge it into `mkDerivation`"?

I would say YAGNI. The options could be deprecated and copied into such an attrset if that is needed one day.
But maybe this can be worked around be overriding the attrs if the `run` derivation? In which case this could be closed if it's too much of an edge case 🤔
(I'll have to look into that later)

> 3. Should `runtimeDeps` be added to the shell as well? In that case, can we re-use `extraPackages`?

Yes, I think so.

sandydoo commented:
@mrcjkb, let's get this merged in. Would you have time to work on this?

Let's re-use `extraPackages` and mark somehow mark that `cargoDeps` is specifically for `nix flake check`. Either `settings.check` or some sort of `flakeCheck` option in `pre-commit.nix.

sandydoo commented:
Fixes #455.
Fixes #452.

mrcjkb commented:
> @mrcjkb, let's get this merged in. Would you have time to work on this?

Sure. I'm sick today, but hopefully I'll be able to pick it up again soon 😄

mrcjkb commented:
@sandydoo I finally got around to this again.

I've updated the PR to reuse `extraPackages` (concatenating them from each hook)
and renamed `rust.cargoDeps` to `rust.check.cargoDeps` to clarify that it's only needed for checks.

It would be nice to have `cargoDeps` per hook, but I don't think it's possible to merge them from multiple hooks.

sandydoo commented on modules/pre-commit.nix:
```diff
@@ -28,6 +30,7 @@ let
     if excludes == [ ] then "^$" else "(${concatStringsSep "|" excludes})";
 
   enabledHooks = filterAttrs (id: value: value.enable) cfg.hooks;
+  enabledExtraPackages = builtins.concatLists (mapAttrsToList (_: value: value.extraPackages) cfg.hooks);
```
```suggestion
  enabledExtraPackages = builtins.concatLists (mapAttrsToList (_: value: value.extraPackages) enabledHooks);
```

sandydoo commented on modules/pre-commit.nix:
```diff
@@ -53,34 +56,41 @@ let
+      src = cfg.rootSrc;
+      buildInputs = [ git ];
+      nativeBuildInputs = enabledExtraPackages
+        ++ (if config.settings.rust.check.cargoDeps != null
+      then [
+        cargoSetupHook
+      ]
+      else [ ]);
```
```suggestion
        ++ lib.optional (config.settings.rust.check.cargoDeps != null) cargoSetupHook;
```

sandydoo commented on modules/hooks.nix:
```diff
@@ -67,10 +67,18 @@ in
-      description = "Path to Cargo.toml";
-      default = null;
+    rust = {
+      check.cargoDeps = mkOption {
```
@domenkozar, need to make a call here.
This is a setting that'll only be used to fix `nix flake check`. Other hooks might need to make their own modifications to the derivation used in flake check. 
Should we prefix them somehow or not?

DaRacci commented:
I've been using this and it seems to be in a good position will this be merged soon?
